Mathlib/Analysis/Convex/BetweenList.lean

```diff
@@ -0,0 +1,317 @@
+/-
+Copyright (c) 2025 Joseph Myers. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joseph Myers
+-/
+import Mathlib.Analysis.Convex.Between
+import Mathlib.Data.List.Triplewise
+
+/-!
+# Betweenness for lists of points.
+
+This file defines notions of lists of points in an affine space being in order on a line.
+
+## Main definitions
+
+* `List.Wbtw R l`: The points in list `l` are weakly in order on a line.
+* `List.Sbtw R l`: The points in list `l` are strictly in order on a line.
+
+-/
+
+
+variable (R : Type*) {V V' P P' : Type*}
+
+open AffineEquiv AffineMap
+
+namespace List
+
+section OrderedRing
+
+variable [OrderedRing R] [AddCommGroup V] [Module R V] [AddTorsor V P]
+variable [AddCommGroup V'] [Module R V'] [AddTorsor V' P']
+
+/-- The points in a list are weakly in that order on a line. -/
+protected def Wbtw (l : List P) : Prop :=
+  l.Triplewise (Wbtw R)
+
+variable {R}
+
+lemma wbtw_cons {p : P} {l : List P} : (p :: l).Wbtw R ↔ l.Pairwise (Wbtw R p) ∧ l.Wbtw R :=
+  triplewise_cons
+
+variable (R)
+
+/-- The points in a list are strictly in that order on a line. -/
+protected def Sbtw (l : List P) : Prop :=
+  l.Wbtw R ∧ l.Pairwise (· ≠ ·)
+
+variable (P)
+
+@[simp] lemma wbtw_nil : ([] : List P).Wbtw R := by
+  simp [List.Wbtw]
+
+@[simp] lemma sbtw_nil : ([] : List P).Sbtw R := by
+  simp [List.Sbtw]
+
+variable {P}
+
+@[simp] lemma wbtw_singleton (p₁ : P) : [p₁].Wbtw R := by
+  simp [List.Wbtw]
+
+@[simp] lemma sbtw_singleton (p₁ : P) : [p₁].Sbtw R := by
+  simp [List.Sbtw]
+
+@[simp] lemma wbtw_pair (p₁ p₂ : P) : [p₁, p₂].Wbtw R := by
+  simp [List.Wbtw]
+
+@[simp] lemma sbtw_pair {p₁ p₂ : P} : [p₁, p₂].Sbtw R ↔ p₁ ≠ p₂ := by
+  simp [List.Sbtw]
+
+variable {R}
+
+@[simp] lemma wbtw_triple {p₁ p₂ p₃ : P} : [p₁, p₂, p₃].Wbtw R ↔ Wbtw R p₁ p₂ p₃ := by
+  simp [List.Wbtw]
+
+@[simp] lemma sbtw_triple {p₁ p₂ p₃ : P} : [p₁, p₂, p₃].Sbtw R ↔ Sbtw R p₁ p₂ p₃ := by
+  simp only [List.Sbtw, wbtw_triple, ne_eq, pairwise_cons, mem_cons, not_mem_nil, or_false,
+    forall_eq_or_imp, forall_eq, IsEmpty.forall_iff, implies_true, Pairwise.nil, and_self, and_true]
+  exact ⟨fun ⟨hw, ⟨h₁₂, h₁₃⟩, h₂₃⟩ ↦ ⟨hw, Ne.symm h₁₂, h₂₃⟩,
+         fun h ↦ ⟨h.1, ⟨h.2.1.symm, h.left_ne_right⟩, h.2.2⟩⟩
+
+lemma wbtw_four {p₁ p₂ p₃ p₄ : P} : [p₁, p₂, p₃, p₄].Wbtw R ↔
+    Wbtw R p₁ p₂ p₃ ∧ Wbtw R p₁ p₂ p₄ ∧ Wbtw R p₁ p₃ p₄ ∧ Wbtw R p₂ p₃ p₄ := by
+  simp [List.Wbtw, triplewise_cons, and_assoc]
+
+lemma sbtw_four {p₁ p₂ p₃ p₄ : P} : [p₁, p₂, p₃, p₄].Sbtw R ↔
+    Sbtw R p₁ p₂ p₃ ∧ Sbtw R p₁ p₂ p₄ ∧ Sbtw R p₁ p₃ p₄ ∧ Sbtw R p₂ p₃ p₄ := by
+  simp [List.Sbtw, List.Wbtw, triplewise_cons, Sbtw]
+  aesop
+
+protected lemma Sbtw.wbtw {l : List P} (h : l.Sbtw R) : l.Wbtw R :=
+  h.1
+
+lemma Sbtw.pairwise_ne {l : List P} (h : l.Sbtw R) : l.Pairwise (· ≠ ·) :=
+  h.2
+
+lemma sbtw_iff_triplewise_and_ne_pair {l : List P} :
+    l.Sbtw R ↔ l.Triplewise (Sbtw R) ∧ ¬ ∃ a, l = [a, a] := by
```

Comment:
Is there a reason to pick the `not exists` spelling instead of `forall not`?

Reply:
I've now changed this to "forall not".

Reply:
(As you can see, that change removed two forwards uses of `not_exists` and introduced two backwards uses.)

Reply:
Ok... thanks for trying!